lib/pgauge_exchange.cu

```diff
@@ -190,73 +111,87 @@ namespace quda {
         }
       }
 
-      GaugeFixUnPackArg<Float, recon> arg(data);
-
       qudaDeviceSynchronize();
       for (int d = 0; d < 4; d++) {
-        if ( !commDimPartitioned(d)) continue;
+        if (!commDimPartitioned(d)) continue;
         comm_start(mh_recv_back[d]);
         comm_start(mh_recv_fwd[d]);
 
-        TuneParam tp;
-        tp.block = make_uint3(128, 1, 1);
-        tp.grid = make_uint3((data.SurfaceCB(d) + tp.block.x - 1) / tp.block.x, 1, 1);
-
-        arg.size = data.SurfaceCB(d);
+        arg.threads.x = U.SurfaceCB(d);
         arg.parity = parity;
         arg.face = d;
         arg.dir = dir;
 
         //extract top face
-        arg.array = reinterpret_cast<complex<Float>*>(send_d[d]); 
-        arg.borderid = X[d] - data.R()[d] - 1;
-        qudaLaunchKernel(Kernel_UnPack<recon, Float, true, decltype(arg)>, tp, GFStream[0], arg);
+        arg.pack = true;
+        arg.array = reinterpret_cast<complex<Float>*>(send_d[d]);
+        arg.borderid = X[d] - U.R()[d] - 1;
+        apply(device::get_stream(0));
 
         //extract bottom
         arg.array = reinterpret_cast<complex<Float>*>(sendg_d[d]);
-        arg.borderid = data.R()[d];
-        qudaLaunchKernel(Kernel_UnPack<recon, Float, true, decltype(arg)>, tp, GFStream[1], arg);
+        arg.borderid = U.R()[d];
+        apply(device::get_stream(1));
 
-        qudaMemcpyAsync(send[d], send_d[d], bytes[d], cudaMemcpyDeviceToHost, GFStream[0]);
-        qudaMemcpyAsync(sendg[d], sendg_d[d], bytes[d], cudaMemcpyDeviceToHost, GFStream[1]);
+        qudaMemcpyAsync(send[d], send_d[d], bytes[d], qudaMemcpyDeviceToHost, device::get_stream(0));
```

Comment:
really makes me wonder whether at some point we should just require a CUDA aware MPI ...

Reply:
Outside of the scope of this PR, but it's my intention to remove this file and just use the regular extended gauge field exchange.  There's some functionality in this exchange though that needs to be ported to the general first though.